### PR TITLE
Migrated IPND changes from bachelor repo to fork of dtn7-rs

### PR DESCRIPTION
**Pull request by IceBlockProduction**

### CODE REVIEW

File: src/ipnd/services.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    /// Creates a new ServiceBlock without any services or clas
+    pub fn new() -> ServiceBlock {
+        ServiceBlock {
+            clas: Vec::new(),
+            services: HashMap::new(),
+        }
+    }
+
+    /// Returns the vector of ConvergencyLayerAgents
+    pub fn get_clas(&self) -> &Vec<(String, Option<u16>)> {
```
Comment (gh0st42) [inline src/ipnd/services.rs:33]: idiomatic rust `get_clas` -> `clas`
https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344]

File: src/ipnd/services.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            convert.push(format!("{}:{}", scheme, port.unwrap_or(0)));
+        }
+        convert
+    }
+
+    pub fn convert_services(&self) -> HashMap<u8, String> {
+        let mut convert: HashMap<u8, String> = HashMap::new();
+        for (tag, payload) in &self.services {
+            match tag {
+                63 => {
```
Comment (gh0st42) [inline src/ipnd/services.rs:50]: these numbers are hard to remember. either constant u8s or enums as u8 should be used here

File: src/ipnd/services.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+                _ => {
+                    info!("Unknown Service encountered. Compare senders IPND version with this one to check for incompatibilities.");
+                }
+            }
+        }
+        convert
+    }
+
+    /// Returns the vector of user defined services
+    pub fn get_services(&self) -> &HashMap<u8, Vec<u8>> {
```
Comment (gh0st42) [inline src/ipnd/services.rs:82]: https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344]

File: src/ipnd/services.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    pub fn add_cla(&mut self, name: &String, port: &Option<u16>) {
+        self.clas.push((name.clone(), port.clone()))
+    }
+
+    /// This method adds a custom service to the HashMap of a ServiceBlock
+    pub fn add_custom_service(&mut self, tag: u8, service: &Vec<u8>) {
+        self.services.insert(tag, service.clone());
+    }
+
+    pub fn set_clas(&mut self, clas: Vec<(String, Option<u16>)>) {
```
Comment (gh0st42) [inline src/ipnd/services.rs:96]: docs missing

File: src/ipnd/services.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        self.clas = clas;
+    }
+
+    pub fn set_services(&mut self, services: HashMap<u8, Vec<u8>>) {
```
Comment (gh0st42) [inline src/ipnd/services.rs:100]: docs missing

File: src/ipnd/services.rs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    ///
+    /// Performs checks on tag and payload combinations
+    ///
+    /// to make sure that the tag and payload content match
+    pub fn build_custom_service(tag: u8, payload: &str) -> Result<(u8, Vec<u8>), String> {
+        match tag {
+            // Reserved tags
+            0..=62 => Err(String::from("Please refrain from using tags between 0 and 62. Those tags are reserved for future enhancements.")),
+            // One tag to allow a random unformatted string
+            63 => {
```
Comment (gh0st42) [inline src/ipnd/services.rs:114]: see comment above, numbers are hard to keep in sync

File: src/ipnd/services.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+                format!("{}. CLA: Name = {}\n", counter, name)
+            };
+            output.push_str(str.as_str());
+            counter += 1;
+        }
+        counter = 0;
+        output.push_str("Other services:\n");
+        for (tag, payload) in self.get_services() {
+            let str = match tag {
+                63 => format!(
```
Comment (gh0st42) [inline src/ipnd/services.rs:192]: same with the numbers as above.. if one changes then it has to be fixed in several places

File: src/lib.rs [hunk 8]
```diff
@@ -34,6 +35,35 @@ lazy_static! {
...
 }
+pub fn service_add(tag: u8, service: String) {
+    (*DTNCORE.lock()).service_list.insert(tag, service);
+}
+pub fn add_discovery_destination(destination: &String) {
+    (*CONFIG.lock())
+        .discovery_destinations
+        .insert(destination.clone(), 0);
+}
+pub fn update_sequence(destination: &String) {
```
Comment (gh0st42) [inline src/lib.rs:46]: whats a sequence in this context? is this the correct term? seems very generic and not ipnd specific .. maybe add docstring or rename a bit

File: tests/beacon-test.rs [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+    println!("Starting deserialization after: {}ns", startd);
+    for x in serialized {
+        serde_cbor::from_slice::<Beacon>(&x).expect("Something bad happened");
+    }
+    let afterd = now.elapsed().unwrap().as_nanos();
+    println!("Deserialization finished after: {}ns", afterd - startd);
+}
+
+#[test]
+#[ignore]
```
Comment (gh0st42) [inline tests/beacon-test.rs:306]: why is this test and the following ignored? please add a comment

File: src/lib.rs (hunk 8 above)
Comment (gh0st42) [reply src/lib.rs:46]: if these are for the beacon sequence number, why isn't this logic handled as part of the discovery_destination? sounds like this bsn stuff should be just methods directly on the destination

File: examples/dtn7.toml.example [hunk 10]
```diff
@@ -1,6 +1,7 @@
 # Example config file for dtn7 daemon
-debug = false
+debug = true
```
Comment (gh0st42) [inline examples/dtn7.toml.example:2]: default should be no debug

**Review by gh0st42 — CHANGES_REQUESTED** [review]
thank you for this PR! 
code looks good but I have some minor issues that should be addressed before merging.

File: tests/beacon-test.rs (hunk 9 above)
Comment (IceBlockProduction) [reply tests/beacon-test.rs:304]: Added comments for why those tests are ignored

**Review by IceBlockProduction — COMMENTED** [review]
Took care of all requested changes in the latest commit but the one about specifying a custom listening port

File: src/bin/dtnd.rs [hunk 11]
```diff
@@ -301,7 +301,7 @@ Tag 255 takes 5 arguments and is interpreted as address. Usage: -S 255:'Samplest
     }
     if let Some(destinations) = matches.values_of("discoverydestination") {
         for destination in destinations {
-            cfg.add_destination(String::from(destination))
+            cfg.add_destination(&mut String::from(destination))
```
Comment (gh0st42) [inline src/bin/dtnd.rs:304]: is it really necessary to pass a mutable string here? does not seem very clean.
From an outside perspective, it is not clear what is happening in add_destination and why something is mutated in there.

File: src/dtnconfig.rs [hunk 12]
```diff
@@ -233,11 +233,21 @@ impl DtnConfig {
...
-        let addr: SocketAddr = destination
-            .parse()
-            .expect("Error: Unable to parse the provided address into IP format");
+
+    /// Helper function that adds discovery destinations to a config struct
+    /// 
+    /// When provided with an IP address without port the default port 3003 is appended
+    pub fn add_destination(&mut self, destination: &mut String) -> std::io::Result<()> {
+        let addr: SocketAddr = if destination.parse::<SocketAddr>().is_err() {
+            destination.push_str(":3003");
```
Comment (gh0st42) [inline src/dtnconfig.rs:242]: as this is the only variant where something is added a non-mutable string as a parameter should be fine.
here `destination` could be locally reassigned to `format!()` to generate a new string with the port. then the function is self-contained and without side effects for the calling parent.

File: src/bin/dtnd.rs [hunk 13]
```diff
@@ -73,7 +73,7 @@ async fn main() -> std::io::Result<()> {
             Arg::with_name("discoverydestination")
                 .short("E")
                 .long("discovery-destination")
-                .value_name("DD:local_port")
+                .value_name("DD[:local_port]")
```
Comment (gh0st42) [inline src/bin/dtnd.rs:76]: actually, this is the `remote_port` isn't it? `local_port` would refer to a port on the local machine, the listening port. 
I guess in this case just `port` would also be fine.

**Review by gh0st42 — CHANGES_REQUESTED** [review]
Some minor points to keep the API clean.
Thank you for your effort!

**Pull request merged**